File: business-logic/classic-scanlog-core/tests/semantic_conformance/registry_keys.rs
//! Public registry key constants shared with Python.
use super::{RunnerResult, invalid};
use classic_registry_core::Keys;
use serde_json::{Value, json};

/// Read every common public key directly; no fixture-authored result enters execution.
pub(super) fn execute(fixture: &Value) -> RunnerResult<Value> {
    if fixture != &json!({"request":{}}) {
        return Err(invalid("unexpected registry keys input").into());
    }
    Ok(
        json!({"keys":{"YAML_CACHE":Keys::YAML_CACHE,"MANUAL_DOCS_GUI":Keys::MANUAL_DOCS_GUI,"GAME_PATH_GUI":Keys::GAME_PATH_GUI,"GAME_PATH":Keys::GAME_PATH,"DOCS_PATH":Keys::DOCS_PATH,"IS_GUI_MODE":Keys::IS_GUI_MODE,"OPEN_FILE_FUNC":Keys::OPEN_FILE_FUNC,"GAME":Keys::GAME,"GAME_VERSION":Keys::GAME_VERSION,"VERSION_AUTO_DETECTED":Keys::VERSION_AUTO_DETECTED,"LOCAL_DIR":Keys::LOCAL_DIR,"IS_PRERELEASE":Keys::IS_PRERELEASE,"XSE_VALID":Keys::XSE_VALID,"XSE_VERSION":Keys::XSE_VERSION,"ENB_PRESENT":Keys::ENB_PRESENT,"GAME_VERSION_DETECTED":Keys::GAME_VERSION_DETECTED}}),
    )
}

File: tools/binding_compliance/conformance/families/registry_keys.py
"""Exact public registry Keys carrier observations."""

import json
from collections.abc import Mapping

from ..coverage import CoveragePredicate, FamilyCoveragePolicy

NAMES = (
    "YAML_CACHE",
    "MANUAL_DOCS_GUI",
    "GAME_PATH_GUI",
    "GAME_PATH",
    "DOCS_PATH",
    "IS_GUI_MODE",
    "OPEN_FILE_FUNC",
    "GAME",
    "GAME_VERSION",
    "VERSION_AUTO_DETECTED",
    "LOCAL_DIR",
    "IS_PRERELEASE",
    "XSE_VALID",
    "XSE_VERSION",
    "ENB_PRESENT",
    "GAME_VERSION_DETECTED",
)


def _observed(value):
    """Require every common public key without granting unnamed constants credit."""
    return (
        isinstance(value, Mapping)
        and set(value) == {"keys"}
        and isinstance(value["keys"], Mapping)
        and set(value["keys"]) == set(NAMES)
        and all(isinstance(v, str) and bool(v) for v in value["keys"].values())
    )


def validate_registry_keys_pack(document, root):
    """Reject inputs containing result values before reading public constants."""
    fixture_root = (root / document["fixtureRoot"]).resolve()
    paths = []
    for case in document["scenarios"]:
        reference = case["input"].get("fixtureRef")
        if (
            case["action"] != "registry-keys.observe"
            or case["input"] != {"fixtureRef": reference}
            or case["fixtureRefs"] != [reference]
        ):
            raise ValueError("registry keys must declare only their fixture")
        path = (fixture_root / document["fixtures"][reference]).resolve()
        if (
            not path.is_relative_to(fixture_root)
            or json.loads(path.read_text()) != {"request": {}}
            or not _observed(case["expected"])
        ):
            raise ValueError("invalid registry keys input or observation")
        paths.append(path)
    return tuple(paths)


REGISTRY_KEYS_COVERAGE_POLICY = FamilyCoveragePolicy(
    "registry-keys",
    (
        CoveragePredicate(
            id="registry-keys-public-constants",
            capability_id="registry-keys.observe",
            action="registry-keys.observe",
            observation_family="key-values",
            rust_symbols=("Keys",),
            runtime_operations=(None,),
            matches=_observed,
        ),
    ),
)
